modules/videoio/src/cap_ffmpeg_impl.hpp

```diff
@@ -88,6 +92,62 @@ extern "C" {
 #include <libavcodec/avcodec.h>
 #include <libswscale/swscale.h>
 
+// https://github.com/FFmpeg/FFmpeg/blob/b6af56c034759b81985f8ea094e41cbd5f7fecfb/doc/APIchanges#L602-L605
+#if LIBAVFORMAT_BUILD < CALC_FFMPEG_VERSION(58, 9, 100)
+#  define CV_FFMPEG_REGISTER
+#endif
+
+// https://github.com/FFmpeg/FFmpeg/blob/b6af56c034759b81985f8ea094e41cbd5f7fecfb/doc/APIchanges#L654-L657
+#if LIBAVCODEC_BUILD < CALC_FFMPEG_VERSION(58, 9, 100)
+#  define CV_FFMPEG_LOCKMGR
+#endif
+
+// https://github.com/FFmpeg/FFmpeg/blob/b6af56c034759b81985f8ea094e41cbd5f7fecfb/doc/APIchanges#L390-L392
+#if LIBAVCODEC_BUILD >= CALC_FFMPEG_VERSION(58, 87, 100)
+#include <libavcodec/bsf.h>
+#endif
+
+// https://github.com/FFmpeg/FFmpeg/blob/b6af56c034759b81985f8ea094e41cbd5f7fecfb/doc/APIchanges#L208-L210
+#if LIBAVFORMAT_BUILD >= CALC_FFMPEG_VERSION(59, 0, 100)
+#  define CV_FFMPEG_FMT_CONST const
+#else
+#  define CV_FFMPEG_FMT_CONST
+#endif
+
+// https://github.com/FFmpeg/FFmpeg/blob/b6af56c034759b81985f8ea094e41cbd5f7fecfb/doc/APIchanges#L623-L624
+#if LIBAVFORMAT_BUILD >= CALC_FFMPEG_VERSION(58, 7, 100)
+#  define CV_FFMPEG_URL
+#endif
+
+// AVStream.codec deprecated in favor of AVStream.codecpar
+// https://github.com/FFmpeg/FFmpeg/blob/b6af56c034759b81985f8ea094e41cbd5f7fecfb/doc/APIchanges#L1039-L1040
+#if LIBAVFORMAT_BUILD >= CALC_FFMPEG_VERSION(59, 16, 100)
```

Comment:
@mshabunin Can you remember why you used 
> #if LIBAVFORMAT_BUILD >= CALC_FFMPEG_VERSION(59, 16, 100)

instead of
> //#if LIBAVFORMAT_BUILD >= CALC_FFMPEG_VERSION(57, 33, 100)

or even better which tests failed?

I am writing a PR to encapsulate raw encoded video using `cv::VideoWriter` to enable `cudacodec::Videowriter` to output video to containers instead of just raw .h264[5], without first initializing the encoder inside `cv::VideoWriter`.  This works perfectly on versions of FFMpeg >= 59.16.100 but on earlier versions after AVStream.codec was depreciated it fails as it is expecting the AVStream.codecpar struct to be used.

Reply:
There is some issues with tests for FFmpeg >=3.1. I've tested it using this repository: https://github.com/mshabunin/videoio-ffmpeg-check I don't remember which tests I had run that time, but at this moment several synthetic and acceleration tests fail:
* Log with condition changed to `>=57.33.100` (FFmpeg >= 3.1) - [log_ffmpeg-3.1.11.txt](https://github.com/opencv/opencv/files/12778016/log_ffmpeg-3.1.11.txt)
* Log with current 4.x - [log_ffmpeg-3.1.11.txt](https://github.com/opencv/opencv/files/12778017/log_ffmpeg-3.1.11.txt)

I decided that this functionality was not mature before FFmpeg 5.0 and thus set condition to `>=59.16.100`. I did not investigate this issue in details.